test/e2e/background-socket/window-handles.ts

```diff
@@ -0,0 +1,208 @@
+import log from 'loglevel';
+import { ThenableWebDriver } from 'selenium-webdriver';
+import { getServerMochaToBackground } from './server-mocha-to-background';
+
+type Handle = {
+  id: string;
+  title: string;
+  url: string;
+};
+
+export type WindowProperties = 'title' | 'url';
+
+/**
+ * Keeps a list of window handles and their properties (title, url).
+ * Takes control of switchToWindowWithTitle and switchToWindowWithUrl away from driver.js
+ */
+export class WindowHandles {
+  driver: ThenableWebDriver;
+
+  rawHandles: string[] = [];
+
+  annotatedHandles: Handle[] = [];
+
+  constructor(driver: ThenableWebDriver) {
+    this.driver = driver;
+  }
+
+  // Gets all window handles and annotates them with title and url
+  async getAllWindowHandles() {
+    this.rawHandles = await this.driver.getAllWindowHandles();
+
+    this.updateAnnotatedHandles();
+
+    log.debug('rawHandles', this.rawHandles);
+    log.debug('annotatedHandles', this.annotatedHandles);
+
+    return this.rawHandles;
+  }
+
+  // Remove outdated annotatedHandles and add new annotatedHandles
+  updateAnnotatedHandles() {
+    // Remove any annotatedHandles that are no longer present
+    for (let i = 0; i < this.annotatedHandles.length; i++) {
+      const handleId = this.annotatedHandles[i].id;
+      if (this.rawHandles.indexOf(handleId) === -1) {
+        log.debug('Removing handle:', this.annotatedHandles[i]);
+        this.annotatedHandles.splice(i, 1);
+        i -= 1;
+      }
+    }
+
+    // Add any new rawHandles to the annotatedHandles
+    for (const handleId of this.rawHandles) {
+      if (!this.annotatedHandles.find((a) => a.id === handleId)) {
+        this.annotatedHandles.push({ id: handleId, title: '', url: '' });
+      }
+    }
+  }
+
+  /**
+   * Update the title and url of the current window.
+   *
+   * @param property - 'title' or 'url'
+   * @param optionalCurrentHandle - If we already know the current handle, we can pass it in here
+   * @returns Depending on `property`, returns the title or url of the current window, or null if no return value is requested
+   */
+  async getCurrentWindowProperties(
+    property?: WindowProperties,
```

Comment:
The description and name are a misaligned

> Update the title and url of the current window.

> getCurrentWindowProperties

Can this be renamed, or maybe split into 2 functions??

Reply:
Whoops, copy/paste mistake, fixed!